File: mcp/tools/mcp-use/libraries/python/mcp_use/agents/adapters/langchain_adapter.py
"""
LangChain adapter for MCP tools.

This module provides utilities to convert MCP tools to LangChain tools.
"""

import re
from typing import Any, NoReturn

from jsonschema_pydantic import jsonschema_to_pydantic
from langchain_core.tools import BaseTool
from mcp.types import (
    AudioContent,
    BlobResourceContents,
    CallToolResult,
    EmbeddedResource,
    ImageContent,
    Prompt,
    ReadResourceRequestParams,
    Resource,
    TextContent,
    TextResourceContents,
)
from mcp.types import (
    Tool as MCPTool,
)
from pydantic import BaseModel, Field, create_model

from mcp_use.agents.adapters.base import BaseAdapter
from mcp_use.client.connectors.base import BaseConnector
from mcp_use.errors.error_formatting import format_error
from mcp_use.logging import logger

LangChainContentBlock = dict[str, Any]
LangChainToolResult = str | LangChainContentBlock | list[LangChainContentBlock]


def _mcp_content_to_langchain(content: list[Any]) -> str | list[LangChainContentBlock]:
    """Convert MCP tool result content to LangChain-compatible format.

    Maps MCP content types to LangChain content blocks:
      - TextContent        → {"type": "text", "text": "..."}
      - ImageContent       → {"type": "image", "source_type": "base64", ...}
      - AudioContent       → {"type": "audio", "source_type": "base64", ...}
      - EmbeddedResource   → text or file block depending on resource type

    If the result is a single TextContent, returns a plain string for simplicity.
    """
    if not content:
        return ""

    # Single TextContent → plain string (most common case)
    if len(content) == 1 and isinstance(content[0], TextContent):
        return content[0].text

    blocks: list[LangChainContentBlock] = []
    for item in content:
        match item:
            case TextContent():
                blocks.append({"type": "text", "text": item.text})
            case ImageContent():
                blocks.append(
                    {
                        "type": "image",
                        "source_type": "base64",
                        "data": item.data,
                        "mime_type": item.mimeType,
                    }
                )
            case AudioContent():
                blocks.append(
                    {
                        "type": "audio",
                        "source_type": "base64",
                        "data": item.data,
                        "mime_type": item.mimeType,
                    }
                )
            case EmbeddedResource():
                resource = item.resource
                if isinstance(resource, TextResourceContents):
                    blocks.append({"type": "text", "text": resource.text})
                elif isinstance(resource, BlobResourceContents):
                    blocks.append(
                        {
                            "type": "file",
                            "source_type": "base64",
                            "data": resource.blob,
                            "mime_type": resource.mimeType or "application/octet-stream",
                        }
                    )
                else:
                    blocks.append({"type": "text", "text": str(resource)})
            case _:
                # Fallback for unknown types
                blocks.append({"type": "text", "text": str(item)})

    if not blocks:
        return ""

    # If all blocks are text, join them as a plain string
    if all(b["type"] == "text" for b in blocks):
        return "\n".join(b["text"] for b in blocks)

    return blocks


class LangChainAdapter(BaseAdapter[BaseTool]):
    """Adapter for converting MCP tools to LangChain tools."""

    framework: str = "langchain"

    def __init__(self, disallowed_tools: list[str] | None = None) -> None:
        """Initialize a new LangChain adapter.

        Args:
            disallowed_tools: list of tool names that should not be available.
        """
        super().__init__(disallowed_tools=disallowed_tools)
        self._connector_tool_map: dict[BaseConnector, list[BaseTool]] = {}
        self._connector_resource_map: dict[BaseConnector, list[BaseTool]] = {}
        self._connector_prompt_map: dict[BaseConnector, list[BaseTool]] = {}

        self.tools: list[BaseTool] = []
        self.resources: list[BaseTool] = []
        self.prompts: list[BaseTool] = []

    def _convert_tool(self, mcp_tool: MCPTool, connector: BaseConnector) -> BaseTool | None:
        """Convert an MCP tool to LangChain's tool format.

        Args:
            mcp_tool: The MCP tool to convert.
            connector: The connector that provides this tool.

        Returns:
            A LangChain BaseTool.
        """
        # Skip disallowed tools
        if mcp_tool.name in self.disallowed_tools:
            return None

        # This is a dynamic class creation, we need to work with the self reference
        adapter_self = self

        class McpToLangChainAdapter(BaseTool):
            name: str = mcp_tool.name or "NO NAME"
            description: str = mcp_tool.description or ""
            # Convert JSON schema to Pydantic model for argument validation
            args_schema: type[BaseModel] = jsonschema_to_pydantic(
                adapter_self.fix_schema(mcp_tool.inputSchema)  # Apply schema conversion
            )
            tool_connector: BaseConnector = connector  # Renamed variable to avoid name conflict
            handle_tool_error: bool = True

            def __repr__(self) -> str:
                return f"MCP tool: {self.name}: {self.description}"

            def _run(self, **kwargs: Any) -> NoReturn:
                """Synchronous run method that always raises an error.

                Raises:
                    NotImplementedError: Always raises this error because MCP tools
                        only support async operations.
                """
                raise NotImplementedError("MCP tools only support async operations")

            async def _arun(self, **kwargs: Any) -> LangChainToolResult:
                """Asynchronously execute the tool with given arguments.

                Args:
                    kwargs: The arguments to pass to the tool.

                Returns:
                    The result of the tool execution.

                Raises:
                    ToolException: If tool execution fails.
                """
                logger.debug(f'MCP tool: "{self.name}" received input: {kwargs}')

                try:
                    tool_result: CallToolResult = await self.tool_connector.call_tool(self.name, kwargs)
                    converted_content: LangChainToolResult | None = None
                    try:
                        converted_content = _mcp_content_to_langchain(tool_result.content)
                        if tool_result.isError:
                            error_message = (
                                converted_content
                                if isinstance(converted_content, str)
                                else "MCP tool returned an error result"
                            )
                            raise RuntimeError(error_message or "MCP tool returned an empty error result")
                        return converted_content
                    except Exception as e:
                        # Log the exception for debugging
                        logger.error(f"Error parsing tool result: {e}")
                        return format_error(
                            e,
                            tool=self.name,
                            tool_content=converted_content if converted_content is not None else tool_result.content,
                        )

                except Exception as e:
                    if self.handle_tool_error:
                        return format_error(e, tool=self.name)  # Format the error to make LLM understand it
                    raise

        return McpToLangChainAdapter()

    def _convert_resource(self, mcp_resource: Resource, connector: BaseConnector) -> BaseTool:
        """Convert an MCP resource to LangChain's tool format.

        Each resource becomes an async tool that returns its content when called.
        The tool takes **no** arguments because the resource URI is fixed.
        """

        def _sanitize(name: str) -> str:
            return re.sub(r"[^A-Za-z0-9_]+", "_", name).lower().strip("_")

        class ResourceTool(BaseTool):
            name: str = _sanitize(mcp_resource.name or f"resource_{mcp_resource.uri}")
            description: str = (
                mcp_resource.description or f"Return the content of the resource located at URI {mcp_resource.uri}."
            )
            args_schema: type[BaseModel] = ReadResourceRequestParams
            tool_connector: BaseConnector = connector
            handle_tool_error: bool = True

            def _run(self, **kwargs: Any) -> NoReturn:
                raise NotImplementedError("Resource tools only support async operations")

            async def _arun(self, **kwargs: Any) -> Any:
                logger.debug(f'Resource tool: "{self.name}" called')
                try:
                    result = await self.tool_connector.read_resource(mcp_resource.uri)
                    for content in result.contents:
                        # Attempt to decode bytes if necessary
                        if isinstance(content, bytes):
                            content_decoded = content.decode()
                        else:
                            content_decoded = str(content)

                    return content_decoded
                except Exception as e:
                    if self.handle_tool_error:
                        return format_error(e, tool=self.name)  # Format the error to make LLM understand it
                    raise

        return ResourceTool()

    def _convert_prompt(self, mcp_prompt: Prompt, connector: BaseConnector) -> BaseTool:
        """Convert an MCP prompt to LangChain's tool format.

        The resulting tool executes `get_prompt` on the connector with the prompt's name and
        the user-provided arguments (if any). The tool returns the decoded prompt content.
        """
        prompt_arguments = mcp_prompt.arguments

        # Sanitize the prompt name to create a valid Python identifier for the model name
        base_model_name = re.sub(r"[^a-zA-Z0-9_]", "_", mcp_prompt.name)
        if not base_model_name or base_model_name[0].isdigit():
            base_model_name = "PromptArgs_" + base_model_name
        dynamic_model_name = f"{base_model_name}_InputSchema"

        if prompt_arguments:
            field_definitions_for_create: dict[str, Any] = {}
            for arg in prompt_arguments:
                param_type: type = getattr(arg, "type", str)
                if arg.required:
                    field_definitions_for_create[arg.name] = (
                        param_type,
                        Field(description=arg.description),
                    )
                else:
                    field_definitions_for_create[arg.name] = (
                        param_type | None,
                        Field(None, description=arg.description),
                    )

            InputSchema = create_model(dynamic_model_name, **field_definitions_for_create, __base__=BaseModel)
        else:
            # Create an empty Pydantic model if there are no arguments
            InputSchema = create_model(dynamic_model_name, __base__=BaseModel)

        class PromptTool(BaseTool):
            name: str = mcp_prompt.name
            description: str | None = mcp_prompt.description

            args_schema: type[BaseModel] = InputSchema
            tool_connector: BaseConnector = connector
            handle_tool_error: bool = True

            def _run(self, **kwargs: Any) -> NoReturn:
                raise NotImplementedError("Prompt tools only support async operations")

            async def _arun(self, **kwargs: Any) -> Any:
                logger.debug(f'Prompt tool: "{self.name}" called with args: {kwargs}')
                try:
                    result = await self.tool_connector.get_prompt(self.name, kwargs)
                    return result.messages
                except Exception as e:
                    if self.handle_tool_error:
                        return format_error(e, tool=self.name)  # Format the error to make LLM understand it
                    raise

        return PromptTool()

File: mcp/tools/mcp-use/libraries/typescript/packages/agent/src/adapters/langchain_adapter.ts
import type { StructuredToolInterface } from "@langchain/core/tools";
import type {
  CallToolResult,
  Tool as MCPTool,
  Resource,
  Prompt,
} from "@modelcontextprotocol/client";
import type { BaseConnector } from "@mcp-use/client";

import { DynamicStructuredTool } from "@langchain/core/tools";
import { z } from "zod";
import { logger } from "@mcp-use/client";
import { BaseAdapter } from "./base.js";

function schemaToZod(schema: unknown): z.ZodType {
  try {
    // MCP tool inputSchema is JSON Schema; Zod 4 converts natively.
    return z.fromJSONSchema(schema as Record<string, unknown>);
  } catch (err) {
    logger.warn(`Failed to convert JSON schema to Zod: ${err}`);
    return z.any();
  }
}

function sanitizeToolName(name: string): string {
  return name
    .replace(/[^A-Za-z0-9_]+/g, "_")
    .toLowerCase()
    .replace(/^_+|_+$/g, "");
}

/** Converts MCP tools, resources, and prompts into LangChain structured tools. */
export class LangChainAdapter extends BaseAdapter<StructuredToolInterface> {
  private usedToolNames: Set<string> = new Set();

  /**
   * @param disallowedTools - MCP tool names to omit during conversion.
   */
  constructor(disallowedTools: string[] = []) {
    super(disallowedTools);
  }

  private reserveName(name: string, kind?: "resource" | "prompt"): string {
    if (!this.usedToolNames.has(name)) {
      this.usedToolNames.add(name);
      return name;
    }
    if (kind) {
      const prefixed = `${kind}_${name}`;
      if (!this.usedToolNames.has(prefixed)) {
        this.usedToolNames.add(prefixed);
        return prefixed;
      }
      // Both base name and prefixed name are taken; fall back to a numeric suffix.
      let i = 2;
      while (this.usedToolNames.has(`${prefixed}_${i}`)) i++;
      const fallback = `${prefixed}_${i}`;
      this.usedToolNames.add(fallback);
      return fallback;
    }
    // No kind: use a numeric suffix to avoid collision.
    let i = 2;
    while (this.usedToolNames.has(`${name}_${i}`)) i++;
    const fallback = `${name}_${i}`;
    this.usedToolNames.add(fallback);
    return fallback;
  }

  /**
   * Converts MCP tools from all connectors and resets name deduplication.
   *
   * @param connectors - Connected MCP connectors.
   * @returns LangChain structured tools.
   */
  public override async createToolsFromConnectors(
    connectors: BaseConnector[]
  ): Promise<StructuredToolInterface[]> {
    // Reset names at the start of each loading cycle.
    this.usedToolNames.clear();
    return super.createToolsFromConnectors(connectors);
  }

  /**
   * Convert a single MCP tool specification into a LangChainJS structured tool.
   */
  protected convertTool(
    mcpTool: MCPTool,
    connector: BaseConnector
  ): StructuredToolInterface | null {
    // Filter out disallowed tools early.
    if (this.disallowedTools.includes(mcpTool.name)) {
      return null;
    }

    // Derive a strict Zod schema for the tool's arguments.
    const argsSchema: z.ZodType = mcpTool.inputSchema
      ? schemaToZod(mcpTool.inputSchema)
      : z.object({}).optional();

    const toolName = this.reserveName(mcpTool.name ?? "NO NAME");
    const tool = new DynamicStructuredTool({
      name: toolName,
      description: mcpTool.description ?? "", // Blank is acceptable but discouraged.
      schema: argsSchema,
      func: async (input: Record<string, any>): Promise<string> => {
        logger.debug(
          `MCP tool "${mcpTool.name}" received input: ${JSON.stringify(input)}`
        );
        try {
          const result: CallToolResult = await connector.callTool(
            mcpTool.name,
            input
          );
          return JSON.stringify(result);
        } catch (err: any) {
          logger.error(`Error executing MCP tool: ${err.message}`);
          return `Error executing MCP tool: ${String(err)}`;
        }
      },
    });

    return tool;
  }

  /**
   * Convert a single MCP resource into a LangChainJS structured tool.
   * Each resource becomes an async tool that returns its content when called.
   */
  protected convertResource(
    mcpResource: Resource,
    connector: BaseConnector
  ): StructuredToolInterface | null {
    const resourceBaseName =
      sanitizeToolName(mcpResource.name || mcpResource.uri) || "resource";
    const resourceName = this.reserveName(resourceBaseName, "resource");
    const resourceUri = mcpResource.uri;

    const tool = new DynamicStructuredTool({
      name: resourceName,
      description:
        mcpResource.description ||
        `Return the content of the resource located at URI ${resourceUri}.`,
      schema: z.object({}).optional(), // Resources take no arguments
      func: async (): Promise<string> => {
        logger.debug(`Resource tool: "${resourceName}" called`);
        try {
          const result = await connector.readResource(resourceUri);
          if (result.contents && result.contents.length > 0) {
            return result.contents
              .map((content: any) => {
                if (typeof content === "string") {
                  return content;
                }
                if (content.text) {
                  return content.text;
                }
                if (content.uri) {
                  return content.uri;
                }
                return JSON.stringify(content);
              })
              .join("\n");
          }
          return "Resource is empty or unavailable";
        } catch (err: any) {
          logger.error(`Error reading resource: ${err.message}`);
          return `Error reading resource: ${String(err)}`;
        }
      },
    });

    return tool;
  }

  /**
   * Convert a single MCP prompt into a LangChainJS structured tool.
   * The resulting tool executes getPrompt on the connector with the prompt's name
   * and the user-provided arguments (if any).
   */
  protected convertPrompt(
    mcpPrompt: Prompt,
    connector: BaseConnector
  ): StructuredToolInterface | null {
    // Build Zod schema from prompt arguments
    let argsSchema: z.ZodType = z.object({}).optional();

    if (mcpPrompt.arguments && mcpPrompt.arguments.length > 0) {
      const schemaFields: Record<string, z.ZodType> = {};
      for (const arg of mcpPrompt.arguments) {
        // All arguments default to string type since type is not available in Prompt definition
        // (Note: MCP spec includes type, but SDK TypeScript types don't)
        const zodType: z.ZodType = z.string();

        if (arg.required !== false) {
          schemaFields[arg.name] = zodType;
        } else {
          schemaFields[arg.name] = zodType.optional();
        }
      }
      argsSchema =
        Object.keys(schemaFields).length > 0
          ? z.object(schemaFields)
          : z.object({}).optional();
    }

    const promptBaseName =
      sanitizeToolName(mcpPrompt.name || "prompt") || "prompt";
    const promptName = this.reserveName(promptBaseName, "prompt");
    const tool = new DynamicStructuredTool({
      name: promptName,
      description: mcpPrompt.description || "",
      schema: argsSchema,
      func: async (input: Record<string, any>): Promise<string> => {
        logger.debug(
          `Prompt tool: "${mcpPrompt.name}" called with args: ${JSON.stringify(input)}`
        );
        try {
          const result = await connector.getPrompt(mcpPrompt.name, input);
          if (result.messages && result.messages.length > 0) {
            return result.messages
              .map((msg: any) => {
                if (typeof msg === "string") {
                  return msg;
                }
                if (msg.content) {
                  return typeof msg.content === "string"
                    ? msg.content
                    : JSON.stringify(msg.content);
                }
                return JSON.stringify(msg);
              })
              .join("\n");
          }
          return "Prompt returned no messages";
        } catch (err: any) {
          logger.error(`Error getting prompt: ${err.message}`);
          return `Error getting prompt: ${String(err)}`;
        }
      },
    });

    return tool;
  }
}
